pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/instanceselector/ReplicaGroupInstanceSelector.java

```diff
@@ -0,0 +1,62 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.v2.instanceselector;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.common.metrics.BrokerMetrics;
+import org.apache.pinot.common.utils.HashUtil;
+
+
+/**
+ * Instance selector for replica-group routing strategy.
+ * <p>The selection algorithm will always evenly distribute the traffic to all replicas of each segment, and will select
+ * the same replica id for all segments with the same number of replicas. The algorithm is very light-weight and will
+ * do best effort to select the least servers for the request.
+ * <p>The algorithm relies on the mirror segment assignment from replica-group segment assignment strategy. With mirror
+ * segment assignment, any server in one replica-group will always have a corresponding server in other replica-groups
+ * that have the same segments assigned. For an example, if S1 is a server in replica-group 1, and it has mirror server
+ * S2 in replica-group 2 and S3 in replica-group 3. All segments assigned to S1 will also be assigned to S2 and S3. In
+ * stable scenario (external view matches ideal state), if we always pick the same replica id for all segments, only
```

Comment:
What is meant by picking up the same replica group id?

If we have 3 replica groups RG1, RG2 and RG3, each with 2 hosts (S1, S2), (S3, S4), (S5, S6) respectively then the problem of evenly distributing the queries boils down to evenly picking up replica groups assuming the mirror segment assignment did the right thing. Correct?

So let's say we have 3 queries (exact same) touching the same number of segment fired concurrently. Can we say that each query will go to a different replica group?

Reply:
There is no guarantee on selecting the servers from the same replica-group, but it is guaranteed that only 2 servers are selected for each query, and each server only need to answer 1 query.
Updated the doc for more explaination